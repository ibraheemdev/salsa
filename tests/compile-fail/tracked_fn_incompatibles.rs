use salsa::Database as Db;

#[salsa::input]
struct MyInput {
    field: u32,
}

#[salsa::tracked(data = Data)]
fn tracked_fn_with_data(db: &dyn Db, input: MyInput) -> u32 {
    input.field(db) * 2
}

#[salsa::tracked(db = Db)]
fn tracked_fn_with_db(db: &dyn Db, input: MyInput) -> u32 {
    input.field(db) * 2
}

#[salsa::tracked(revisions = 12)]
fn tracked_fn_with_revisions(db: &dyn Db, input: MyInput) -> u32 {
    input.field(db) * 2
}

#[salsa::tracked(constructor = TrackedFn3)]
fn tracked_fn_with_constructor(db: &dyn Db, input: MyInput) -> u32 {
    input.field(db) * 2
}

#[salsa::tracked]
fn tracked_fn_with_one_input(db: &dyn Db) -> u32 {}

#[salsa::tracked]
fn tracked_fn_with_receiver_not_applied_to_impl_block(&self, db: &dyn Db) -> u32 {}

#[salsa::tracked(specify)]
fn tracked_fn_with_too_many_arguments_for_specify(
    db: &dyn Db,
    input: MyInput,
    input: MyInput,
) -> u32 {
}

#[salsa::interned]
struct MyInterned<'db> {
    field: u32,
}

#[salsa::tracked]
fn tracked_fn_with_lt_param_and_elided_lt_on_db_arg1<'db>(
    db: &dyn Db,
    interned: MyInterned<'db>,
) -> u32 {
    interned.field(db) * 2
}

#[salsa::tracked]
fn tracked_fn_with_lt_param_and_elided_lt_on_db_arg2<'db_lifetime>(
    db: &dyn Db,
    interned: MyInterned<'db_lifetime>,
) -> u32 {
    interned.field(db) * 2
}

#[salsa::tracked]
fn tracked_fn_with_lt_param_and_elided_lt_on_input<'db>(
    db: &'db dyn Db,
    interned: MyInterned,
) -> u32 {
    interned.field(db) * 2
}

#[salsa::tracked]
fn tracked_fn_with_multiple_lts<'db1, 'db2>(db: &'db1 dyn Db, interned: MyInterned<'db2>) -> u32 {
    interned.field(db) * 2
}

fn main() {}
