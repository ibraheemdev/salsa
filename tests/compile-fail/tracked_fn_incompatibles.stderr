error: `data` option not allowed here
 --> tests/compile-fail/tracked_fn_incompatibles.rs:8:18
  |
8 | #[salsa::tracked(data = Data)]
  |                  ^^^^

error: `db` option not allowed here
  --> tests/compile-fail/tracked_fn_incompatibles.rs:13:18
   |
13 | #[salsa::tracked(db = Db)]
   |                  ^^

error: `revisions` option not allowed here
  --> tests/compile-fail/tracked_fn_incompatibles.rs:18:18
   |
18 | #[salsa::tracked(revisions = 12)]
   |                  ^^^^^^^^^

error: `constructor` option not allowed here
  --> tests/compile-fail/tracked_fn_incompatibles.rs:23:18
   |
23 | #[salsa::tracked(constructor = TrackedFn3)]
   |                  ^^^^^^^^^^^

error: #[salsa::tracked] must also be applied to the impl block for tracked methods
  --> tests/compile-fail/tracked_fn_incompatibles.rs:32:55
   |
32 | fn tracked_fn_with_receiver_not_applied_to_impl_block(&self, db: &dyn Db) -> u32 {}
   |                                                       ^^^^^

error: only functions with a single salsa struct as their input can be specified
  --> tests/compile-fail/tracked_fn_incompatibles.rs:34:18
   |
34 | #[salsa::tracked(specify)]
   |                  ^^^^^^^

error: must have a `'db` lifetime
  --> tests/compile-fail/tracked_fn_incompatibles.rs:49:9
   |
49 |     db: &dyn Db,
   |         ^

error: must have a `'db_lifetime` lifetime
  --> tests/compile-fail/tracked_fn_incompatibles.rs:57:9
   |
57 |     db: &dyn Db,
   |         ^

error: only a single lifetime parameter is accepted
  --> tests/compile-fail/tracked_fn_incompatibles.rs:72:39
   |
72 | fn tracked_fn_with_multiple_lts<'db1, 'db2>(db: &'db1 dyn Db, interned: MyInterned<'db2>) -> u32 {
   |                                       ^^^^

error: `self` parameter is only allowed in associated functions
  --> tests/compile-fail/tracked_fn_incompatibles.rs:32:55
   |
32 | fn tracked_fn_with_receiver_not_applied_to_impl_block(&self, db: &dyn Db) -> u32 {}
   |                                                       ^^^^^ not semantically valid as function parameter
   |
   = note: associated functions are those in `impl` or `trait` definitions

error[E0415]: identifier `input` is bound more than once in this parameter list
  --> tests/compile-fail/tracked_fn_incompatibles.rs:38:5
   |
38 |     input: MyInput,
   |     ^^^^^ used as parameter more than once

error[E0106]: missing lifetime specifier
  --> tests/compile-fail/tracked_fn_incompatibles.rs:66:15
   |
66 |     interned: MyInterned,
   |               ^^^^^^^^^^ expected named lifetime parameter
   |
help: consider using the `'db` lifetime
   |
66 |     interned: MyInterned<'db>,
   |                         +++++

error[E0308]: mismatched types
  --> tests/compile-fail/tracked_fn_incompatibles.rs:29:46
   |
28 | #[salsa::tracked]
   | ----------------- implicitly returns `()` as its body has no tail or `return` expression
29 | fn tracked_fn_with_one_input(db: &dyn Db) -> u32 {}
   |                                              ^^^ expected `u32`, found `()`

error[E0308]: mismatched types
  --> tests/compile-fail/tracked_fn_incompatibles.rs:32:78
   |
32 | fn tracked_fn_with_receiver_not_applied_to_impl_block(&self, db: &dyn Db) -> u32 {}
   |    --------------------------------------------------                        ^^^ expected `u32`, found `()`
   |    |
   |    implicitly returns `()` as its body has no tail or `return` expression

error[E0308]: mismatched types
  --> tests/compile-fail/tracked_fn_incompatibles.rs:39:6
   |
35 | fn tracked_fn_with_too_many_arguments_for_specify(
   |    ---------------------------------------------- implicitly returns `()` as its body has no tail or `return` expression
...
39 | ) -> u32 {
   |      ^^^ expected `u32`, found `()`
